tensorflow/core/kernels/collective_ops.cc

```diff
@@ -1281,5 +1281,80 @@ REGISTER_KERNEL_BUILDER(Name("CollectiveAllToAllV3").Device(DEVICE_CPU),
                         CollectiveAllToAllV3OpKernel);
 REGISTER_KERNEL_BUILDER(Name("CollectiveAllToAllV3").Device(DEVICE_GPU),
                         CollectiveAllToAllV3OpKernel);
+
+class CollectiveReduceScatterV2OpKernel : public CollectiveOpV2Kernel {
+ public:
+  explicit CollectiveReduceScatterV2OpKernel(OpKernelConstruction* c)
+      : CollectiveOpV2Kernel(c) {
+    string merge_op_name;
+    OP_REQUIRES_OK(c, c->GetAttr("merge_op", &merge_op_name));
+    if (merge_op_name == "Max") {
+      merge_op_name = "Maximum";
+    } else if (merge_op_name == "Min") {
+      merge_op_name = "Minimum";
+    }
+    string final_op_name;
+    OP_REQUIRES_OK(c, c->GetAttr("final_op", &final_op_name));
+    OP_REQUIRES_OK(
+        c, c->GetAttr("max_subdivs_per_device", &max_subdivs_per_device_));
+    // Prepare OpKernels for reduction and final operations.
+    // The merge_op takes two inputs
+    NodeDef sub_node;
+    sub_node.add_input(c->def().input(0));
+    sub_node.add_input(c->def().input(0));
+    sub_node.set_device(c->def().device());
+    SetAttrValue(data_type_, &(*sub_node.mutable_attr())["T"]);
+    merge_op_ = BuildOpKernel(c, merge_op_name, &sub_node);
+    final_op_ = BuildOpKernel(c, final_op_name, &sub_node);
+    name_ = strings::StrCat(c->def().name(), ": ReduceScatterV2(", merge_op_name, ",",
+                            final_op_name, ")");
+    VLOG(2) << "CollectiveReduceScatterV2 " << this << " name " << name_
+            << " communication_hint " << communication_hint_;
+  }
+
+  void ComputeAsync(OpKernelContext* c, DoneCallback done) override {
+    auto col_params = new CollectiveParams();
+    auto done_with_cleanup = [col_params, done = std::move(done)]() {
+      done();
+      col_params->Unref();
+    };
+    OP_REQUIRES_OK_ASYNC(c,
+                         FillCollectiveParams(col_params, REDUCE_SCATTER_COLLECTIVE,
+                                              /*group_size*/ c->input(1),
+                                              /*group_key*/ c->input(2),
+                                              /*instance_key*/ c->input(3)),
+                         done_with_cleanup);
+    auto output_shape = c->input(0).shape();
+    output_shape.set_dim(
+        0, output_shape.dim_size(0) / col_params->group.group_size);
+    col_params->instance.shape = output_shape;
+    col_params->merge_op = merge_op_.get();
+    col_params->final_op = final_op_.get();
+    VLOG(1) << "CollectiveReduceScatterV2 group_size " << col_params->group.group_size
+            << " group_key " << col_params->group.group_key << " instance_key "
+            << col_params->instance.instance_key;
+    // Allocate the output tensor, trying to reuse the input.
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK_ASYNC(
+        c, c->allocate_output(0, col_params->instance.shape, &output),
+        done_with_cleanup);
+    Run(c, col_params, std::move(done_with_cleanup));
+  }
+
+ private:
+  int max_subdivs_per_device_;
```

Comment:
Should this attribute be passed to CollectiveParams->instance->CollImplDetails->max_subdivs_per_device?

I realize that currently collective reduce v2 is also not passing this attribute, which we might want to fix.

Reply:
Thanks @ishark for the review!

Looks like this attribute is only used in the ring reduce algorithm for `CollectiveReduceV2`. I guess it could be used when we add a non-nccl implementation for `CollectiveReduceScatterV2` but I'm not 100% sure. Do you think it is needed?